Automation naming and UX updates

PClmnt opened the pull request:
## Description
This PR adds the ability to name automations as well as some other UX style updates around automations

Naming an automation:

https://github.com/Budibase/budibase/assets/5665926/7dfc7809-aa47-4217-b8e9-a3a8b444d2f3

Error and reset states when adding illegal characters and removing a given name

https://github.com/Budibase/budibase/assets/5665926/21617046-dfba-4c56-84be-b3cce77551d4

There are multiple states around this, the name can also only be alphanumeric which we check for on blur. If the name does not satisfy this requirement, and the user moves away from the input then it is reset to the original step name, either a name or the step type. 

The new name is also shown in the bindings drawer:
<img width="321" alt="image" src="https://github.com/Budibase/budibase/assets/5665926/c9cab098-49e3-47c6-9cec-c42af31b5048">

Some of the styling updates include
- Updating the fields to have proper labels instead to the left of the fields
- Moving the action buttons to the top of the block instead of in the body, and updating tooltips and icons around them
- Updating the undo / redo / run test context menu to be sticky like the data section query view

Other fixes:
- Removes Users table from Create Delete row options: #BUDI-7714
- Automation bindings not working correctly for JS scripting: #11866 



<img width="637" alt="image" src="https://github.com/Budibase/budibase/assets/5665926/3b6620a1-a6b4-4101-b669-2ffd798fad0c">

https://github.com/Budibase/budibase/assets/5665926/a1c3f514-adc6-4b1c-b17e-e5829b436b4a


## Feature branch env
[Feature Branch Link](http://fb-feat-automation-naming-ux-updates.fb.qa.budibase.net)

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowChart.svelte:
```diff
@@ -20,7 +14,8 @@
   let confirmDeleteDialog
-
+  let scrolling = false
+  let hasScrolled = false
```
Looking at the usages of `hasScrolled`, I think it's redundant. We can just swap out any usages of it for `scrolling` and get the same behaviour so we may as well remove it.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -1,27 +1,39 @@
+  $: automationName = stepNames[block.id] || block.name || ""
+  $: automationNameError = getAutomationNameError(automationName)
+  $: status = updateStatus(testResult, isTrigger)
+  $: isTrigger = isTrigger || block.type === "TRIGGER"
```
This is a recursive declaration, so I'm not sure how svelte will handle this. I think we'd be better using a new variable name here to distinguish between the prop and this, and it would also help keep things more understandable.

I can see that you just moved this, but still think it's worth addressing since we're using it in other ways now.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -43,10 +55,40 @@
+    if (name.length > 0) {
+      let invalidRoleName = !validRegex.test(name)
+      if (invalidRoleName) {
+        return "Please enter a role name consisting of only alphanumeric symbols and underscores"
```
The copy here mentions "role name" but this has nothing to do with roles does it? We could just use "name" instead. Maybe just a wee leftover from copying and pasting!

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -1,27 +1,39 @@
   }
-  $: isTrigger = isTrigger || block.type === "TRIGGER"
-  $: status = updateStatus(testResult, isTrigger)
+  $: loopBlock = $selectedAutomation?.definition.steps.find(
```
`selectedAutomation` has optional chaining here, but up above we're referencing a nested property without it. I think this should be consistent - it either can be null (in which case we need to update the `stepNames` declaration above) or it can't be null (in which case we don't need optional chaining).

I reckon just try deleting the selected automation and see if it throws an error or not - then we'll know whether this gets rendered one final time when the selected automation is null.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -69,12 +111,32 @@
+            automationName = e.target.value
+            if (!automationNameError) {
+              automationNameError = false // Reset the error when input is valid
+            }
```
This statement doesn't do anything - it's checking if something is falsey, then setting it to false if it is. I think we can just remove this if statement.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -69,12 +111,32 @@
+          autocomplete="off"
+          value={automationName}
+          on:input={e => {
+            automationNameError = getAutomationNameError(e.target.value)
```
`automationNameError` is automatically derived using the reactive statement above, so we don't need to set it here at all. We can just set `automationName` to the new value, and let the reactive statement run to derive any error.

aptkingston replied:
We also might want to trim the value here, because in `saveName` we depend on checking the length of the name, and we wouldn't want to store a blank name that's just a space for example.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -69,12 +111,32 @@
+            typing = false
+            if (automationNameError) {
+              automationName = stepNames[block.id]
+              automationNameError = null
```
We don't need to worry about clearing the error here, as setting `automationName` back to a valid name will automatically clear it for us using the reactive statement above.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -123,4 +215,47 @@
+    box-sizing: border-box;
+    overflow: hidden;
+    text-overflow: ellipsis;
+    white-space: nowrap;
```
It might be nice to add top-left and top-right border radius to the input, to match the actual header, so that it doesn't get squared edges when typing.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -123,4 +215,47 @@
+  }
+
+  .typing {
+    border: 0.5px solid var(--spectrum-global-color-static-blue-500);
```
We should use 1px here rather than 0.5px to be consistent with the rest of the platform.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -123,4 +215,47 @@
+  }
+
+  .typing-error {
+    border: 0.5px solid var(--spectrum-global-color-static-red-500);
```
Same as above, we should use 1px here.

aptkingston commented on packages/builder/src/components/automation/AutomationBuilder/FlowChart/FlowItemHeader.svelte:
```diff
@@ -123,4 +215,47 @@
+    font-family: var(--font-sans);
+    color: var(--ink);
+    background-color: transparent;
+    border: none;
```
We should use a 1px solid transparent border here. This will mean that whenever we start typing, and show a blue border, that the input doesn't jump because its size changes all of a sudden. It'll just make the experience a little smoother for the user.

aptkingston reviewed (commented):
Nice work addressing the feedback!

2 little things I still found:
- When focusing to rename, there's still a jump. Maybe there's still some border differences? Would be great if there was no jump when focusing.
- If you take the default name, like after loading the page, then rename it and use invalid characters on your first attempt, it will revert to the previous value because the current one is invalid. The issue is that when you have no previous value, it will actually insert the string `undefined` as the name, so we should just check if we have a previous value or not, and use the default name if we don't have a previous one to revert to.

PClmnt commented:
> Nice work addressing the feedback!
> 
> 2 little things I still found:
> 
> * When focusing to rename, there's still a jump. Maybe there's still some border differences? Would be great if there was no jump when focusing.
> * If you take the default name, like after loading the page, then rename it and use invalid characters on your first attempt, it will revert to the previous value because the current one is invalid. The issue is that when you have no previous value, it will actually insert the string `undefined` as the name, so we should just check if we have a previous value or not, and use the default name if we don't have a previous one to revert to.

Sorted those two issues, you were right for the border one, was just missing a transparent border on the outer div, seemed to sort it!

mike12345567 commented:
One issue I found with the add test data modal: 
![image](https://github.com/Budibase/budibase/assets/4407001/c20650cf-9dce-4d6f-9d2b-95e58c4f63af)

Unsure if this is related - but was an issue I saw when testing, probably worth solving!

mike12345567 reviewed (approved):
LGTM! If @aptkingston approves then I'm happy to merge - left one comment about something we should probably solve, appears to be the case on feature branch as well, so isn't must my environment!

PClmnt commented:
> LGTM! If @aptkingston approves then I'm happy to merge - left one comment about something we should probably solve, appears to be the case on feature branch as well, so isn't must my environment!

Yea fixed that! Just needed to update the modal sizing

aptkingston commented:
One final issue I found is that you're able to rename multiple steps to the same name. That means that the bindings get confused because the readable to runtime map can't uniquely identify the readable bindings. Here's a quick example where I have 2 steps called `a`:
![image](https://github.com/Budibase/budibase/assets/9075550/83c10786-da55-4ca9-ad53-7a9ceff85cfc)

My backend log step doesn't work as expected because both bindings for `{{ a.rows }}` end up being the same value, even though I'm trying to get the outputs from different steps.

Basically I think we just need to enforce step name uniqueness 👍

PClmnt commented:
> That means that the bindings get confused because the readable to runtime map can't uniquely identify the readable bindings

Sorted, uniqueness is now enforced when adding automation names
